### PR TITLE
[12.0][FIX] product: fix migration of attribute lines

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Solves https://github.com/OCA/OpenUpgrade/issues/1875.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product/migrations/12.0.1.2/pre-migration.py [hunk 1]
```diff
@@ -24,13 +27,27 @@
 _table_renames = [
     ('product_attribute_line', 'product_template_attribute_line'),
     ('product_attribute_price', 'product_template_attribute_value'),
+    ('product_attribute_line_product_attribute_value_rel',
+     'product_attribute_value_product_template_attribute_line_rel'),
 ]
 
 
+def avoid_new_constraint_in_product_template_attribute_line(cr):
+    openupgrade.logged_query(
```
Comment (pedrobaeza) [inline addons/product/migrations/12.0.1.2/pre-migration.py:37]: Why this?

Comment (MiquelRForgeFlow) [reply addons/product/migrations/12.0.1.2/pre-migration.py:37]: In v11 exists this contrain:
```python
        if any(line.value_ids > line.attribute_id.value_ids for line in self):
            raise ValidationError(_('Error ! You cannot use this attribute with the following value.'))
```
In v12, it has changed to:
```python
        if any(not line.value_ids or line.value_ids > line.attribute_id.value_ids for line in self):
            raise ValidationError(_('You cannot use this attribute with the following value.'))
```

It means that you can only create template attribute lines if and only if they are linked to an attribute value or more. You need to drop those lines that are not linked to any attribute value.

Comment (pedrobaeza) [reply addons/product/migrations/12.0.1.2/pre-migration.py:37]: Yeah, but you are destructing information with this, so we shouldn't unless it blocks something. In this case, I see it's a Python constraint, so it's not too much problem to let it.

Comment (MiquelRForgeFlow) [reply addons/product/migrations/12.0.1.2/pre-migration.py:37]: You are not destructing "information" in this case, it's just cleaning unused and useless data. Lines without an attribute value are pointless, that's why odoo added that new constrain.

Comment (pedrobaeza) [reply addons/product/migrations/12.0.1.2/pre-migration.py:37]: OK, got it

File: addons/product/migrations/12.0.1.2/end-migration.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+from openupgradelib import openupgrade
+
+
+def update_product_template_attribute_value(env):
+    # the product.template.attribute.values are recomputed each time a
+    # template.attribute.line changes. The conditions to generate this
+    # template_attribute_value have changed, so we need to recompute them.
+    lines = env['product.template.attribute.line'].search([])
+    lines._update_product_template_attribute_values()
```
Comment (pedrobaeza) [inline addons/product/migrations/12.0.1.2/end-migration.py]: Can we reconstruct only when needed (when different conditions are met)?

Comment (MiquelRForgeFlow) [reply addons/product/migrations/12.0.1.2/end-migration.py]: I will check if we can filter them to just the needed cases.

Comment (MiquelRForgeFlow) [reply addons/product/migrations/12.0.1.2/end-migration.py]: filtered!

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**